src/main/scala/is/hail/sparkextras/ContextRDD.scala

```diff
@@ -0,0 +1,159 @@
+package is.hail.sparkextras
+
+import is.hail.io._
+import is.hail.expr.types._
+import is.hail.utils._
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.storage._
+import scala.reflect.ClassTag
+
+object ContextRDD {
+  def empty[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.emptyRDD[C => Iterator[T]], mkc)
+
+  // this one weird trick permits the caller to specify T without C
+  sealed trait Empty[T] {
+    def apply[C <: AutoCloseable](
+      sc: SparkContext,
+      mkc: () => C
+    )(implicit tct: ClassTag[T]
+    ): ContextRDD[C, T] = empty[C, T](sc, mkc)
+  }
+  private[this] object emptyInstance extends Empty[Nothing]
+  def empty[T] = emptyInstance.asInstanceOf[Empty[T]]
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]]
+  ): ContextRDD[C, T] = {
+    assert(xs.length > 1)
+    val mkc = xs(0).mkc
+    assert(xs.tail.forall(_ == mkc))
+    union(sc, xs, mkc)
+  }
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.union(xs.map(_.rdd)), mkc)
+
+  def weaken[C <: AutoCloseable, T: ClassTag](
+    rdd: RDD[T],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(rdd.mapPartitions(it => Iterator.single((ctx: C) => it)), mkc)
+
+  type ElementType[C, T] = C => Iterator[T]
+}
+
+class ContextRDD[C <: AutoCloseable, T: ClassTag](
+  val rdd: RDD[C => Iterator[T]],
+  val mkc: () => C
+) extends Serializable {
+  def run: RDD[T] =
+    rdd.mapPartitions { part => using(mkc()) { cc => part.flatMap(_(cc)) } }
+
+  private[this] def inCtx[U: ClassTag](
+    f: C => Iterator[U]
+  ): Iterator[C => Iterator[U]] = Iterator.single { (ctx: C) => f(ctx) }
```

Comment:
Why the lambda?  This is just `Iterator.single(f)`.

Reply:
Also, I don't know why it is called `inCtx`, `f` is already expecting a context.

Reply:
I nix'ed the eta expansion.

Basically, I found myself typing out:

```scala
Iterator.single { (ctx: C) => ... }
```

a few times. This method is saving the word `Iterator.single` and the type annotation. I thought it read nicely where it's used:

```scala
rdd.mapPartitions(
  part => inCtx(ctx => f(ctx, part.flatMap(_(ctx)))),
  preservesPartitioning),
mkc)
```

I'm mapping over the partitions, and, for each partition, I'm doing some computation in the given context.